fpga_interchange/arch_pack_clusters.cc

```diff
@@ -370,6 +506,344 @@ static bool check_cluster_cells_compatibility(CellInfo *old_cell, CellInfo *new_
+                }
+                copy_idx_to_cells[arc.first].insert(cell);
+                std::queue<std::pair<uint32_t, uint32_t>> workqueue;
+                while(!prep.empty()){
+                    uint32_t idx = prep.front(); prep.pop();
+                    for (const auto &connection : cluster->connection_graph[idx].connections)
+                        if (arc.first != connection.target_idx)
+                            workqueue.push(std::pair<uint32_t, uint32_t>(arc.first, connection.target_idx));
+                }
+                binary_constraint_check(cluster, workqueue, copy_idx_to_cells, ctx);
+                if (back_solver(cluster, copy_idx_to_cells, ctx)){
+                    idx_to_cells = std::move(copy_idx_to_cells);
+                    return true;
+                }
+            }
+        }
+    }
+    return true;
+}
+
+void Arch::prepare_macro_cluster( const ClusterPOD *cluster, uint32_t index)
+{
+    Context *ctx = getCtx();
+    IdString cluster_name(cluster->name);
+
+    pool<IdString> cluster_cell_types;
+    for (auto cell_type : cluster->root_cell_types)
+        cluster_cell_types.insert(IdString(cell_type));
+
+    // Find cluster roots for each macro only ones
+    dict<IdString, CellInfo *> roots;
+    for (auto &cell : cells){
+        CellInfo *ci = cell.second.get();
+        if(ci->macro_parent == IdString())
+            continue;
+        if(ci->cluster != ClusterId())
+            continue;
+        if (!cluster_cell_types.count(ci->type))
+            continue;
+        if(roots.count(ci->macro_parent))
+            continue;
+        // Simple check based on cell type counting
+        dict<IdString, uint32_t> cells_in_macro, counter;
+        pool<IdString> cell_types;
+        for (auto &cell_type : cluster->required_cells){
+            cells_in_macro[IdString(cell_type.name)] = cell_type.count;
+            cell_types.insert(IdString(cell_type.name));
+        }
+
+        for (auto &node_cell : macro_to_cells[ci->macro_parent]){
+            auto cell_type = node_cell->type;
+            if(!counter.count(cell_type))
+                counter[cell_type] = 0;
+            counter[cell_type]++;
+            cell_types.insert(cell_type);
+        }
+        bool failed = false;
+        for(auto cell_type : cell_types){
+            if(ctx->verbose && cells_in_macro.count(cell_type))
+                log_info("Required: %s %d\n", cell_type.c_str(ctx), cells_in_macro[cell_type]);
+            if(ctx->verbose && cells_in_macro.count(cell_type))
+                log_info("Have: %s %d\n", cell_type.c_str(ctx), counter[cell_type]);
+            if(!cells_in_macro.count(cell_type) || !counter.count(cell_type) || cells_in_macro[cell_type] != counter[cell_type])
+                failed = true;
+            if(failed && ctx->verbose)
+                log_info("Cell count stage failed, for sure not this cluster\n");
+            if(failed)
+                break;
+        }
+        if(failed){
```

Comment:
I think this needs a `make clangformat` run on it

Reply:
Run make clangformat on project